File: python_app/models/transaction.py
"""Transaction Record model - translated from TRNREC.cpy copybook.

Mirrors the COBOL TRANSACTION-RECORD structure with composite key
(date, time, portfolio_id, sequence_no) and transaction data fields.
"""

from datetime import datetime
from decimal import Decimal
from enum import StrEnum

from pydantic import BaseModel, Field


class TransactionType(StrEnum):
    """Transaction type codes from 88-level values in TRNREC."""

    BUY = "BU"
    SELL = "SL"
    TRANSFER = "TR"
    FEE = "FE"


class TransactionStatus(StrEnum):
    """Transaction status codes from 88-level values in TRNREC."""

    PENDING = "P"
    DONE = "D"
    FAILED = "F"
    REVERSED = "R"


class TransactionKey(BaseModel):
    """Composite key for transaction record (TRN-KEY)."""

    date: str = Field(max_length=8, description="Transaction date YYYYMMDD")
    time: str = Field(max_length=6, description="Transaction time HHMMSS")
    portfolio_id: str = Field(max_length=8, description="Portfolio identifier")
    sequence_no: str = Field(max_length=6, description="Sequence number")


class TransactionRecord(BaseModel):
    """Full transaction record translated from COBOL TRANSACTION-RECORD.

    Maps to TRNREC.cpy copybook fields:
    - TRN-KEY (composite key)
    - TRN-DATA (transaction details)
    - TRN-AUDIT (audit trail)
    """

    # Key fields (TRN-KEY)
    date: str = Field(max_length=8, description="TRN-DATE: YYYYMMDD")
    time: str = Field(max_length=6, description="TRN-TIME: HHMMSS")
    portfolio_id: str = Field(max_length=8, description="TRN-PORTFOLIO-ID")
    sequence_no: str = Field(max_length=6, description="TRN-SEQUENCE-NO")

    # Data fields (TRN-DATA)
    investment_id: str = Field(max_length=10, description="TRN-INVESTMENT-ID")
    type: TransactionType = Field(description="TRN-TYPE: BU/SL/TR/FE")
    quantity: Decimal = Field(max_digits=15, decimal_places=4, description="TRN-QUANTITY S9(11)V9(4)")
    price: Decimal = Field(max_digits=15, decimal_places=4, description="TRN-PRICE S9(11)V9(4)")
    amount: Decimal = Field(max_digits=15, decimal_places=2, description="TRN-AMOUNT S9(13)V9(2)")
    currency: str = Field(max_length=3, description="TRN-CURRENCY")
    status: TransactionStatus = Field(description="TRN-STATUS: P/D/F/R")

    # Audit fields (TRN-AUDIT)
    process_date: str = Field(default="", max_length=26, description="TRN-PROCESS-DATE")
    process_user: str = Field(default="", max_length=8, description="TRN-PROCESS-USER")

    @property
    def composite_key(self) -> str:
        """Build the composite key matching VSAM KSDS key structure."""
        return f"{self.date}{self.time}{self.portfolio_id}{self.sequence_no}"

    def mark_done(self, user: str) -> None:
        """Mark transaction as done with audit trail."""
        self.status = TransactionStatus.DONE
        self.process_date = datetime.now().isoformat()
        self.process_user = user

    def mark_failed(self, user: str) -> None:
        """Mark transaction as failed with audit trail."""
        self.status = TransactionStatus.FAILED
        self.process_date = datetime.now().isoformat()
        self.process_user = user

File: modern/src/types/transaction.ts
/**
 * Transaction types and Zod schemas
 * Translated from: src/copybook/common/TRNREC.cpy
 *
 * Transaction types: BU=Buy, SL=Sell, TR=Transfer, FE=Fee
 * Status: P=Pending, D=Done, F=Failed, R=Reversed
 */

import { z } from "zod";

// ---------------------------------------------------------------------------
// Enums (from COBOL Level 88 conditionals)
// ---------------------------------------------------------------------------

export const TransactionType = {
  Buy: "BU",
  Sell: "SL",
  Transfer: "TR",
  Fee: "FE",
} as const;
export type TransactionType =
  (typeof TransactionType)[keyof typeof TransactionType];

export const TransactionStatus = {
  Pending: "P",
  Done: "D",
  Failed: "F",
  Reversed: "R",
} as const;
export type TransactionStatus =
  (typeof TransactionStatus)[keyof typeof TransactionStatus];

// ---------------------------------------------------------------------------
// Zod schemas
// ---------------------------------------------------------------------------

export const transactionTypeSchema = z.enum(["BU", "SL", "TR", "FE"]);

export const transactionStatusSchema = z.enum(["P", "D", "F", "R"]);

/**
 * Maps to TRN-KEY group
 * TRN-DATE (YYYYMMDD) + TRN-TIME (HHMMSS) + TRN-PORTFOLIO-ID + TRN-SEQUENCE-NO
 */
export const transactionKeySchema = z.object({
  date: z.string().length(8),
  time: z.string().length(6),
  portfolioId: z.string().length(8),
  sequenceNo: z.string().length(6),
});

/** Maps to TRN-DATA group */
export const transactionDataSchema = z.object({
  investmentId: z.string().length(10),
  type: transactionTypeSchema,
  quantity: z.number(),
  price: z.number(),
  amount: z.number(),
  currency: z.string().length(3),
  status: transactionStatusSchema,
});

/** Maps to TRN-AUDIT group */
export const transactionAuditSchema = z.object({
  processDate: z.string().max(26),
  processUser: z.string().max(8),
});

/** Full TRANSACTION-RECORD from TRNREC.cpy */
export const transactionRecordSchema = z.object({
  key: transactionKeySchema,
  data: transactionDataSchema,
  audit: transactionAuditSchema,
});

// ---------------------------------------------------------------------------
// TypeScript interfaces
// ---------------------------------------------------------------------------

export interface TransactionKey {
  date: string;
  time: string;
  portfolioId: string;
  sequenceNo: string;
}

export interface TransactionData {
  investmentId: string;
  type: TransactionType;
  quantity: number;
  price: number;
  amount: number;
  currency: string;
  status: TransactionStatus;
}

export interface TransactionAudit {
  processDate: string;
  processUser: string;
}

export interface TransactionRecord {
  key: TransactionKey;
  data: TransactionData;
  audit: TransactionAudit;
}
